File: _examples/java/example-java-camunda-data/src/test/java/io/holunda/addons/camunda/data/CamundaDataAddonTest.java
package io.holunda.addons.camunda.data;

import org.jetbrains.annotations.NotNull;
import org.junit.Test;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class CamundaDataAddonTest {

    private static final Variable<String> FOO = new Variable<String>() {
        @Override public String getKey() {
            return "foo";
        }

        @NotNull @Override public Class<String> getType() {
            return String.class;
        }
    };

    private static final Variable<String> BAR = new Variable<String>() {
        @Override public String getKey() {
            return "bar";
        }

        @NotNull @Override public Class<String> getType() {
            return String.class;
        }
    };

    @Test
    public void reader() {
        Map<String, Object> variables = Collections.singletonMap("foo", "hello");

        MyVariableReaderWriter reader = new MyVariableReaderWriter(variables);

        assertThat(reader.readRequired(FOO)).isEqualTo("hello");

        assertThatThrownBy(() -> reader.readRequired(BAR))
                .isInstanceOf(VariableNotFoundException.class);
    }

    @Test
    public void writer() {
        Map<String, Object> variables = new HashMap<>();
        MyVariableReaderWriter writer = new MyVariableReaderWriter(variables);

        writer.write(FOO, "world");

        assertThat(writer.readRequired(FOO)).isEqualTo("world");
    }

    public static class MyVariableReaderWriter implements VariableReader, VariableWriter {

        private final Map<String, Object> store;

        public MyVariableReaderWriter(Map<String, Object> store) {
            this.store = store;
        }

        @Override public <T> Optional<T> read(Variable<T> variable) {
            return (Optional<T>) Optional.ofNullable(store.get(variable.getKey()));
        }

        @NotNull @Override public <T> T readRequired(@NotNull Variable<T> variable) {
            return read(variable).orElseThrow(() -> new VariableNotFoundException(variable));
        }

        @NotNull @Override public <T> VariableWriter write(@NotNull Variable<T> variable, @NotNull T value) {
            store.put(variable.getKey(), value);
            return this;
        }
    }
}

File: _examples/kotlin/example-kotlin-camunda-data/src/test/kotlin/CamundaDataAddonTest.kt
package io.holunda.addons.camunda.data

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.Test
import java.util.*

class CamundaDataAddonTest {
    companion object {
        val FOO = object : Variable<String> {
            override val key: String
                get() = "foo"
            override val type: Class<String>
                get() = String::class.java
        }

        val BAR = object : Variable<String> {
            override val key: String
                get() = "bar"
            override val type: Class<String>
                get() = String::class.java
        }


    }

    @Test
    fun `can implement reader interface`() {
        val reader = MyVariableReader(mutableMapOf(FOO.key to "hello"))

        assertThat(reader.readRequired(FOO)).isEqualTo("hello")
        assertThatThrownBy { reader.readRequired(BAR) }.isInstanceOf(VariableNotFoundException::class.java)
    }

    @Test
    fun `implement writer interface`() {
        val writer = MyVariableReader(mutableMapOf())

        writer.write(FOO, "world")

        assertThat(writer.readRequired(FOO)).isEqualTo("world")
    }

    class MyVariableReader(val store: MutableMap<String, Any>) : VariableReader, VariableWriter {
        override fun <T : Any> read(variable: Variable<T>): Optional<T> = Optional.ofNullable(store[variable.key] as T?);

        override fun <T : Any> write(variable: Variable<T>, value: T): MyVariableReader {
            store.put(variable.key, value)
            return this
        }

    }
}